File: ios/Shared/Data/FriendlyError.swift
import Foundation

/// Maps raw errors to one-line, user-appropriate messages. Technical detail
/// stays in os_log; the UI never shows HTTP bodies or server internals.
extension Error {
    var zenUserMessage: String {
        if let urlError = self as? URLError, Self.isOffline(urlError.code) {
            return String(localized: "error.offline")
        }
        let ns = self as NSError
        if ns.domain == NSURLErrorDomain, Self.isOffline(URLError.Code(rawValue: ns.code)) {
            return String(localized: "error.offline")
        }
        if let sync = self as? SyncError {
            switch sync {
            case .totpRequired, .notSignedIn:
                return sync.localizedDescription
            case .storageUnavailable:
                return String(localized: "error.storage_unavailable")
            case .conflict:
                return String(localized: "error.conflict")
            case .auth:
                return String(localized: "error.auth")
            case .crypto:
                return String(localized: "error.crypto")
            case .network:
                return String(localized: "error.network")
            }
        }
        return String(localized: "error.generic")
    }

    private static func isOffline(_ code: URLError.Code) -> Bool {
        switch code {
        case .notConnectedToInternet, .networkConnectionLost, .timedOut,
             .cannotFindHost, .dnsLookupFailed, .dataNotAllowed:
            true
        default:
            false
        }
    }
}

File: android/app/src/main/java/de/kjell/zencompanion/util/FriendlyError.kt
package de.kjell.zencompanion.util

import androidx.annotation.StringRes
import de.kjell.zencompanion.R
import de.kjell.zencompanion.sync.SyncError
import java.io.IOException
import java.net.ConnectException
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.net.ssl.SSLException

/**
 * Port of `Shared/FriendlyError.swift`: maps raw errors to one-line,
 * user-appropriate messages. The UI never shows HTTP bodies.
 */
object FriendlyError {
    @StringRes
    fun messageRes(error: Throwable): Int = when (error) {
        is SyncError.NotSignedIn -> R.string.error_not_signed_in
        is SyncError.TotpRequired -> R.string.error_totp
        is SyncError.StorageUnavailable -> R.string.error_storage_unavailable
        is SyncError.Auth -> R.string.error_auth
        is SyncError.Crypto -> R.string.error_crypto
        is SyncError.Conflict -> R.string.error_conflict
        is SyncError.Network -> if (isOffline(error)) R.string.error_offline else R.string.error_network
        else -> if (isOffline(error)) R.string.error_offline else R.string.error_generic
    }

    fun isOffline(error: Throwable): Boolean = when (error) {
        is UnknownHostException,
        is ConnectException,
        is SocketTimeoutException,
        is SocketException,
        is SSLException,
        -> true
        is IOException -> error.message?.let { msg ->
            listOf(
                "offline", "network", "connection", "timeout", "timed out",
                "econn", "etimedout", "ehostunreach", "enetunreach",
            ).any { msg.contains(it, ignoreCase = true) }
        } ?: false
        else -> false
    }
}
